src/librpmem/rpmem.c

```diff
@@ -34,18 +34,243 @@
  * rpmem.c -- main source file for librpmem
  */
 #include <stdlib.h>
+#include <netdb.h>
+#include <stdio.h>
+#include <pthread.h>
+#include <errno.h>
 
 #include "librpmem.h"
+#include "out.h"
+#include "util.h"
+#include "rpmem_common.h"
+#include "rpmem_util.h"
+#include "rpmem_obc.h"
+#include "rpmem_fip.h"
+#include "rpmem_fip_common.h"
 
 /*
- * struct rpmem_pool -- remote pool context
+ * rpmem_pool -- remote pool context
  */
 struct rpmem_pool {
-	void *pool_addr;
-	size_t pool_size;
+	void *pool_addr;		/* local pool address */
+	size_t pool_size;		/* pool size */
+	struct rpmem_obc *obc;		/* out-of-band connection handle */
+	struct rpmem_fip *fip;		/* fabric provider handle */
+	char *user;
+	char *node;
+	char *service;
+	char fip_service[NI_MAXSERV];
+	enum rpmem_provider provider;
+	pthread_t monitor;
+	volatile int error;
 };
 
 /*
+ * rpmem_get_provider -- returns provider based on node address and environment
+ */
+static enum rpmem_provider
+rpmem_get_provider(const char *node)
+{
+	struct rpmem_fip_probe probe;
+
+	int ret;
+
+	enum rpmem_provider prov = RPMEM_PROV_UNKNOWN;
+
+	ret = rpmem_fip_probe_get(node, &probe);
+	if (ret)
+		return prov;
+
+	/*
+	 * The sockets provider can be used only if specified environment
+	 * variable is set to 1.
+	 */
+	if (rpmem_fip_probe(probe, RPMEM_PROV_LIBFABRIC_SOCKETS)) {
+		char *enable_sock_env = getenv(RPMEM_PROV_SOCKET_ENV);
+		if (enable_sock_env) {
+			int enable_sock = atoi(enable_sock_env);
+			if (enable_sock == 1)
+				prov = RPMEM_PROV_LIBFABRIC_SOCKETS;
+		}
+	}
+
+	/* the verbs provider has higher priority */
+	if (rpmem_fip_probe(probe, RPMEM_PROV_LIBFABRIC_VERBS)) {
+		char *enable_verbs_env = getenv(RPMEM_PROV_VERBS_ENV);
+		if (enable_verbs_env) {
+			int enable_verbs = atoi(enable_verbs_env);
+			if (enable_verbs == 1)
+				prov = RPMEM_PROV_LIBFABRIC_VERBS;
+		} else {
+			prov = RPMEM_PROV_LIBFABRIC_VERBS;
+		}
+	}
+
+	return prov;
```

Comment:
rpmem_get_provider() does not set errno if (prov == RPMEM_PROV_UNKNOWN):
rpmem0.log <librpmem>: <1> [rpmem.c:160 rpmem_common_init] cannot find provider: Success